### PR TITLE
Set `--lib es6`, and use `Map`/`Set` methods where possible.

### PR DESCRIPTION
**Pull request by andy-hanson**
#### PR checklist

- [X] Addresses an existing issue: #1971
- [ ] New feature, bugfix, or enhancement
  - [ ] Includes tests
- [ ] Documentation update
- [X] Enable CircleCI for your fork (https://circleci.com/add-projects)

#### What changes did you make?

Set the `"lib"` option to `["es6", "dom"]` ("dom" typings used by underscore).
**EDIT**: Removed underscore; it's only used by 1 function, `camelize`, that's a one-liner. [Their source](https://github.com/epeli/underscore.string/blob/master/camelize.js), [mine](https://github.com/palantir/tslint/pull/1984/files#diff-e1495d267619047a7cca5cfe8f692729R46).

Used `Map` and `Set` in cases where their absence appeared to have been compensated for.

#### Is there anything you'd like reviewers to focus on?

It's more efficient to lookup in a Map only once, so avoid this pattern:
```ts
if (map.has(key)) {
	use(map.get(key))
}
```

Prefer:
```ts
const got = map.get(key);
if (got !== undefined) {
	use(got);
}
```

### CODE REVIEW

File: src/rules/noMagicNumbersRule.ts [hunk 1]
```diff
@@ -68,25 +68,21 @@ export class Rule extends Lint.Rules.AbstractRule {
...
 
     public visitNode(node: ts.Node) {
         const isUnary = this.isUnaryNumericExpression(node);
-        if (node.kind === ts.SyntaxKind.NumericLiteral && node.parent !== undefined && !Rule.ALLOWED_NODES[node.parent.kind] || isUnary) {
-            const text = node.getText();
-            if (!this.allowed[text]) {
-                this.addFailureAtNode(node, Rule.FAILURE_STRING);
-            }
+        const isNumber = node.kind === ts.SyntaxKind.NumericLiteral && !Rule.ALLOWED_NODES.has(node.parent!.kind);
+        const isMagicNumber = (isNumber || isUnary) && !this.allowed.has(node.getText());
```
Comment (andy-hanson) [inline src/rules/noMagicNumbersRule.ts]: Seems like the `isUnary` case should also check for `ALLOWED_NODES`, but this is what it did before.

File: src/rules/preferForOfRule.ts [hunk 2]
```diff
@@ -47,17 +47,15 @@ interface IIncrementorState {
...
 // a map of incrementors and whether or not they are only used to index into an array reference in the for loop
-interface IncrementorMap {
-    [name: string]: IIncrementorState;
-}
+type IncrementorMap = Map<string, IIncrementorState>;
 
 class PreferForOfWalker extends Lint.BlockScopeAwareRuleWalker<{}, IncrementorMap> {
     public createScope() {
-        return {};
+        return new Map();
```
Comment (ajafff) [inline src/rules/preferForOfRule.ts]: this method could just return `void` (need to update type parameters). `scope` is never used anywhere, only `blockScope` is needed.

File: src/language/walker/skippableTokenAwareRuleWalker.ts [hunk 3]
```diff
@@ -21,11 +21,10 @@ import {IOptions} from "../rule/rule";
 import {RuleWalker} from "./ruleWalker";
 
 export class SkippableTokenAwareRuleWalker extends RuleWalker {
-    protected tokensToSkipStartEndMap: {[start: number]: number};
+    private tokensToSkipStartEndMap = new Map<number, number>();
 
     constructor(sourceFile: ts.SourceFile, options: IOptions) {
```
Comment (nchen63) [inline src/language/walker/skippableTokenAwareRuleWalker.ts]: don't need constructor anymore

File: src/rules/noMagicNumbersRule.ts (hunk 1 above)
Comment (nchen63) [inline src/rules/noMagicNumbersRule.ts]: Agree. Opened issue #2005

**Comment by adidahiya** [comment]
this is an awesome PR, thanks @andy-hanson

**Pull request merged**